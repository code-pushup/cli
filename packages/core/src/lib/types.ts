export type GlobalOptions = {
  // Show progress bar in stdout
  progress: boolean;
  // Outputs additional information for a run
  verbose: boolean;
};
