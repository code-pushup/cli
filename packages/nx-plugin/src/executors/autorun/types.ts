import { GeneralExecutorOnlyOptions } from '../internal/types';

export type AutorunExecutorOnlyOptions = {
  projectPrefix: string;
} & GeneralExecutorOnlyOptions;
