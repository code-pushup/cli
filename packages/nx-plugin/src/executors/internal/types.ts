import { ProjectConfiguration } from 'nx/src/config/workspace-json-project-json';

/**
 * Types used in the executor only
 */
export type GeneralExecutorOnlyOptions = {
  dryRun?: boolean;
};

/**
 * executor types that apply for a subset of exector's.
 * In this case the project related options
 *
 */
export type ProjectExecutorOnlyOptions = {
  projectPrefix?: string;
};

/**
 * CLI types that apply globally for all commands.
 */
export type GlobalExecutorOptions = {
  verbose?: boolean;
  progress?: boolean;
  config?: string;
};

/**
 * CLI types that apply for a subset of commands.
 * In this case the collection of data (collect, autorun, history)
 */
export type CollectExecutorOnlyOptions = {
  onlyPlugins?: string[];
};

/**
 * context that is normalized for all executor's
 */
export type BaseNormalizedExecutorContext = {
  projectConfig?: ProjectConfiguration;
} & { workspaceRoot: string };
