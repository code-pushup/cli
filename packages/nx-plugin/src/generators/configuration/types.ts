import type { ItemOrArray } from '@code-pushup/utils';

export type ExecutableCode = {
  fileImports: ItemOrArray<string>;
  codeStrings: ItemOrArray<string>;
};
