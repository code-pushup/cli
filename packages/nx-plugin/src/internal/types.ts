export type DynamicTargetOptions = {
  // @TODO add prefix https://github.com/code-pushup/cli/issues/619
  targetName?: string;
};
