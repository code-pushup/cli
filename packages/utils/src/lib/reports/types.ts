import {
  AuditReport,
  CategoryConfig,
  Group,
  PluginReport,
  Report,
} from '@code-pushup/models';

export type ScoredCategoryConfig = CategoryConfig & { score: number };

export type ScoredGroup = Group & {
  score: number;
};

export type ScoredReport = Omit<Report, 'plugins' | 'categories'> & {
  plugins: (Omit<PluginReport, 'groups'> & {
    groups?: ScoredGroup[];
  })[];
  categories: ScoredCategoryConfig[];
};

export type SortableGroup = ScoredGroup & {
  weight: number;
  plugin: string;
};

export type SortableAuditReport = AuditReport & {
  weight: number;
  plugin: string;
};

export type DiffOutcome = 'positive' | 'negative' | 'neutral' | 'mixed';
