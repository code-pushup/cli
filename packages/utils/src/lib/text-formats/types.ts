// eslint-disable-next-line no-magic-numbers
export type Hierarchy = 1 | 2 | 3 | 4 | 5 | 6;
