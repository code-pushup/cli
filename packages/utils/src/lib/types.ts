export type ExcludeNullFromPropertyTypes<T> = {
  [P in keyof T]: Exclude<T[P], null>;
};
