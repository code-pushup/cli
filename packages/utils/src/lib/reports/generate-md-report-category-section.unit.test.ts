import { describe, expect, it } from 'vitest';
import {
  categoriesDetailsSection,
  categoriesOverviewSection,
  categoryGroupItem,
  categoryRef,
} from './generate-md-report-categoy-section';
import { ScoredGroup, ScoredReport } from './types';

// === Categories Overview Section

describe('categoriesOverviewSection', () => {
  it('should skip categories table if categories are empty', () => {
    const md = categoriesOverviewSection({
      plugins: [],
      categories: [],
    });
    expect(md).toBe('');
  });

  it('should render complete categories table', () => {
    expect(
      categoriesOverviewSection({
        plugins: [
          {
            slug: 'eslint',
            title: 'Eslint',
          },
          {
            slug: 'lighthouse',
            title: 'Lighthouse',
          },
        ],
        categories: [
          {
            slug: 'bug-prevention',
            title: 'Bug Prevention',
            score: 0.98,
            refs: [{ slug: 'no-let', type: 'audit' }],
          },
          {
            slug: 'performance',
            title: 'Performance',
            score: 0.74,
            refs: [{ slug: 'largest-contentful-paint', type: 'audit' }],
          },
          {
            slug: 'typescript',
            title: 'Typescript',
            score: 0.14,
            refs: [{ slug: 'no-any', type: 'audit' }],
          },
        ],
      } as ScoredReport),
    ).toMatchSnapshot();
  });
});

// === Categories Details

describe('categoryRef', () => {
  it('should render partial category reference', () => {
    expect(
      categoryRef(
        {
          slug: 'score-report-performance',
          title: 'Score Report Performance',
          value: 12_245,
          score: 1,
        },
        'lighthouse',
      ),
    ).toBe(
      '- ðŸŸ© [Score Report Performance](#score-report-performance-lighthouse) (_lighthouse_) - **12245**',
    );
  });

  it('should render complete category reference', () => {
    expect(
      categoryRef(
        {
          slug: 'score-report-performance',
          title: 'Score Report Performance',
          value: 1,
          score: 0,
          displayValue: '12 errors',
        },
        'lighthouse',
      ),
    ).toBe(
      '- ðŸŸ¥ [Score Report Performance](#score-report-performance-lighthouse) (_lighthouse_) - **12 errors**',
    );
  });
});

describe('categoryGroupItem', () => {
  it('should render partial category reference', () => {
    expect(
      categoryGroupItem(
        {
          slug: 'bug-prevention',
          title: 'Bug Prevention',
          score: 0.9,
        } as ScoredGroup,
        [
          { title: 'No let', slug: 'no-let', score: 0, value: 23 },
          { title: 'No any', slug: 'no-any', score: 0.6, value: 91 },
        ],
        'Eslint',
      ),
    ).toMatchSnapshot();
  });

  it('should render complete category reference', () => {
    expect(
      categoryGroupItem(
        {
          slug: 'bug-prevention',
          title: 'Bug Prevention',
          score: 0.6,
        } as ScoredGroup,
        [
          {
            title: 'No any',
            slug: 'no-any',
            score: 0,
            value: 12,
            displayValue: '12 errors',
          },
          { title: 'No let', slug: 'no-let', score: 1, value: 0 },
        ],
        'Eslint',
      ),
    ).toMatchSnapshot();
  });
});

describe('categoriesDetails', () => {
  it('should render complete categories details', () => {
    expect(
      categoriesDetailsSection({
        plugins: [
          {
            slug: 'eslint',
            title: 'Eslint',
            audits: [
              { slug: 'no-let', title: 'No let', score: 0, value: 5 },
              { slug: 'no-any', title: 'No any', score: 1, value: 0 },
            ],
          },
          {
            slug: 'lighthouse',
            title: 'Lighthouse',
            audits: [
              {
                slug: 'largest-contentful-paint',
                title: 'Largest Contentful Paint',
                score: 0.7,
                value: 2905,
              },
            ],
          },
        ],
        categories: [
          {
            slug: 'bug-prevention',
            title: 'Bug Prevention',
            score: 0.98,
            refs: [{ slug: 'no-let', type: 'audit', plugin: 'eslint' }],
          },
          {
            slug: 'performance',
            title: 'Performance',
            score: 0.74,
            refs: [
              {
                slug: 'largest-contentful-paint',
                type: 'audit',
                plugin: 'lighthouse',
              },
            ],
          },
          {
            slug: 'typescript',
            title: 'Typescript',
            score: 0.14,
            refs: [{ slug: 'no-any', type: 'audit', plugin: 'eslint' }],
          },
        ],
      } as ScoredReport),
    ).toMatchSnapshot();
  });
});
