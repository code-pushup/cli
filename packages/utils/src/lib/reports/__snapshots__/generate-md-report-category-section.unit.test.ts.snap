// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`categoriesDetails > should render complete categories details 1`] = `
"## 🏷 Categories

### Bug Prevention

🟢 Score:  **98**
- 🟥 [No let](#no-let-eslint) (_Eslint_) - **5**

### Performance

🟡 Score:  **74**
- 🟨 [Largest Contentful Paint](#largest-contentful-paint-lighthouse) (_Lighthouse_) - **2905**

### Typescript

🔴 Score:  **14**
- 🟩 [No any](#no-any-eslint) (_Eslint_) - **0**
"
`;

exports[`categoriesOverviewSection > should render complete categories table 1`] = `
"|🏷 Category|⭐ Score|🛡 Audits|
|:--|:--:|:--:|
|[Bug Prevention](#bug-prevention)|🟢 **98**|1|
|[Performance](#performance)|🟡 **74**|1|
|[Typescript](#typescript)|🔴 **14**|1|
"
`;

exports[`categoryGroupItem > should render complete category reference 1`] = `
"- 🟡 Bug Prevention (_Eslint_)
  - 🟥 [No any](#no-any-eslint) - **12 errors**
  - 🟩 [No let](#no-let-eslint) - **0**"
`;

exports[`categoryGroupItem > should render partial category reference 1`] = `
"- 🟢 Bug Prevention (_Eslint_)
  - 🟥 [No let](#no-let-eslint) - **23**
  - 🟨 [No any](#no-any-eslint) - **91**"
`;
